File: PUNet/PacketBuilder.py
from typing import Dict, List
import struct

import yaml
import os

from PUNet.Packet import Packet


class PacketBuilder:
    def __init__(self, packet_directory: str) -> None:
        self.packet_directory = packet_directory

        # header: {field: type}
        self.packet_schemes: Dict[str, Dict[str, chr]] = {}
        self.packet_schemes = self._get_conf_dict()

    def build_from_header(self, header: str) -> Packet:
        """Builds an empty packet with only empty fields.

        :param header: the header of the packet
        :type header: str
        :return: the empty packet
        :rtype: Packet
        """
        return Packet(header, self.format_of(header), *self.fields_of(header))

    def build_from_raw(self, header: str, raw: bytes) -> Packet:
        """Builds a packet from raw bytes.

        :param header: the header of the packet
        :type header: str
        :param raw: the raw bytes of the packet body
        :type raw: bytes
        :return: the built packet
        :rtype: Packet
        """
        new_pack = self.build_from_header(header)
        fields = dict(zip(self.fields_of(header),
                      struct.unpack(">" + self.format_of(header), raw)))
        new_pack.set_fields(**fields)
        return new_pack

    def size_of(self, header: str) -> int:
        """Returns the number of bytes the packet has.

        :param header: the header of the packet
        :type header: str
        :return: the number of bytes the packet has
        :rtype: int
        """
        return struct.calcsize(self.format_of(header))

    def format_of(self, header: str) -> str:
        """Returns the format of the packet struct.

        :param header: the header of the packet
        :type header: str
        :return: the format of the packet
        :rtype: str
        """
        return "".join(self.packet_schemes[header].values())

    def fields_of(self, header: str) -> List[str]:
        """Returns the fields of the packet.

        :param header: the header of the packet
        :type header: str
        :return: the fields of the packet
        :rtype: List[str]
        """
        return self.packet_schemes[header].keys()

    def _get_conf_dict(self) -> dict[str, dict[str, chr]]:
        """Returns a dictionary of all packet's stractures from the
         parsed packet configs.

        :return: a dictionary of all packet's stractures
        :rtype: dict[str, dict[str, chr]]
        """
        general_dict = {}  # A dictionary of all packet's stractures.
        # Go over all of the files and connectes thier packet structures into
        # one dictionary.
        for filename in os.listdir(self.packet_directory):
            if filename.endswith(".packet"):
                conf_yaml = open(self.packet_directory + '\\' + filename, "r")
                yaml_data = yaml.load(conf_yaml, Loader=yaml.FullLoader)
                conf_yaml.close()
                general_dict.update(yaml_data)

        return general_dict

File: RIONet/ga2230net/src/main/java/org/ga2230net/packets/PacketBuilder.java
package org.ga2230net.packets;

import org.yaml.snakeyaml.Yaml;
import java.io.InputStream;
import java.util.LinkedHashMap;
import java.util.Map;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;

import org.ga2230net.socket_utils.StructUtils;

/**
 * Used to build packets according to their config files
 * A packet should only be created using a PacketBuilder
 */
public class PacketBuilder {
    // header to fields and types
    private final Map<String, Map<String, Character>> packetSchemes;

    /**
     * Creates a new packet builder from a packet directory
     * @param packet_directory the packet directory
     */
    public PacketBuilder(String packet_directory) {
        packetSchemes = parsePacketDirectory(packet_directory);
    }

    /**
     * Builds an empty packet from a header
     * @param header the header of the packet
     * @return the packet
     * @throws IllegalArgumentException if the header does not exist in the packet directory
     */
    public Packet buildFromHeader(String header) throws IllegalArgumentException {
        if (!packetSchemes.containsKey(header)) {
            throw new IllegalArgumentException("Header " + header + " does not exist in packet directory");
        }
        return new Packet(header, formatOf(header), fieldsOf(header));
    }

    /**
     * Builds a packet from a header and its raw body
     * @param header the header of the packet
     * @param raw_body the raw body of the packet
     * @return the packet
     * @throws IllegalArgumentException if the header does not exist in the packet directory
     */
    public Packet buildFromRaw(String header, byte[] raw_body) throws IllegalArgumentException {
        if (!packetSchemes.containsKey(header)) {
            throw new IllegalArgumentException("Header " + header + " does not exist in packet directory");
        }
        Packet new_packet = buildFromHeader(header);
        Object[] body = StructUtils.unpack(formatOf(header), raw_body);
        for (int i = 0; i < body.length; i++) {
            new_packet.setField(fieldsOf(header)[i], body[i]);
        }
        return new_packet;
    }

    /**
     * Gets the number of bytes a packet will be from its header
     * @param header the header of the packet
     * @return the size of the packet
     * @throws IllegalArgumentException if the header does not exist in the packet directory
     */
    public int sizeOf(String header) throws IllegalArgumentException {
        if (!packetSchemes.containsKey(header)) {
            throw new IllegalArgumentException("Header " + header + " does not exist in packet directory");
        }
        return StructUtils.sizeOf(formatOf(header));
    }

    /**
     * Gets the struct format of a packet from its header
     * @param header the header of the packet
     * @return the format of the packet
     * @throws IllegalArgumentException if the header does not exist in the packet directory
     */
    public String formatOf(String header) throws IllegalArgumentException {
        if (!packetSchemes.containsKey(header)) {
            throw new IllegalArgumentException("Header " + header + " does not exist in packet directory");
        }
        Map<String, Character> fields = packetSchemes.get(header);
        StringBuilder format = new StringBuilder();
        for (Character type : fields.values()) {
            format.append(type);
        }
        return format.toString();
    }

    /**
     * Gets the fields of a packet from its header
     * @param header the header of the packet
     * @return the fields of the packet
     * @throws IllegalArgumentException if the header does not exist in the packet directory
     */
    public String[] fieldsOf(String header) throws IllegalArgumentException {
        if (!packetSchemes.containsKey(header)) {
            throw new IllegalArgumentException("Header " + header + " does not exist in packet directory");
        }
        return packetSchemes.get(header).keySet().toArray(new String[0]);
    }

    /**
     * Parses a packet directory into a HashMap of headers to fields and types
     * @param packet_directory the packet directory
     * @return the parsed packet directory
     */
    private Map<String, Map<String, Character>> parsePacketDirectory(String packet_directory) {
        // Create an instance of the SnakeYAML library
        Yaml yaml = new Yaml();
        // Create a new HashMap to store the result
        Map<String, Map<String, Character>> result = new LinkedHashMap<>();

        try {
            // Create a new File object representing the directory passed as a parameter
            File folder = new File(packet_directory);
            // Check if the directory exists and if it is a directory
            if(!folder.exists() || !folder.isDirectory()){
                throw new FileNotFoundException("The directory " + packet_directory + " was not found or it's not a directory.");
            }
            // Get an array of all the files in the directory
            File[] listOfFiles = folder.listFiles();
            // Check if the list of files is null
            if (listOfFiles == null) {
                throw new FileNotFoundException("The directory " + packet_directory + " was not found.");
            }
            // Iterate through the files
            for (File file : listOfFiles) {
                // Check if the file is a file and has the ".packet" extension
                if (file.isFile() && file.getName().endsWith(".packet")) {
                    // Read the file using FileInputStream
                    InputStream inputStream = new FileInputStream(file);
                    // Parse the file using the load method of the Yaml class
                    Map<String, Object> map = yaml.load(inputStream);

                    // Iterate through the key-value pairs in the parsed file
                    for (Map.Entry<String, Object> entry : map.entrySet()) {
                        String key = entry.getKey();
                        Map<String, Character> innerMap = new LinkedHashMap<>();
                        for (Map.Entry<String, Object> innerEntry : ((Map<String, Object>) entry.getValue()).entrySet()) {
                            innerMap.put(innerEntry.getKey(), ((String) innerEntry.getValue()).charAt(0));
                        }
                        // Add the key-value pairs to the result map
                        result.put(key, innerMap);
                    }
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        // return the result map
        System.out.println(result);
        return result;
    }
}
